dataflow/run_template/main_test.py

```diff
@@ -44,29 +46,72 @@ def app():
     return flask.Flask(__name__)
 
 
-def unique_job_name(label):
-    return datetime.now().strftime('{}-%Y%m%d-%H%M%S-{}'.format(
-        label, uuid.uuid4().hex))
+# fixture used to create user-friendly names with UUIDs for Dataflow jobs
+# indirect parametrization inspired by this post https://stackoverflow.com/questions/42228895/how-to-parametrize-a-pytest-fixture
+@pytest.fixture(scope="function")
+def dataflow_job_name(request):
+    label = request.param
+    job_name = datetime.now().strftime('{}-%Y%m%d-%H%M%S-{}'.format(
+        label, uuid.uuid4().hex[:5]))
+
+    yield job_name
+
+    # cancel the Dataflow job after running the test
+    dataflow_jobs_cancel(job_name)
+
+
+# Takes in a Dataflow job name and returns its job ID
+def get_job_id_from_name(job_name):
+    # list the 100 most recent Dataflow jobs
+    jobs_request = dataflow.projects().jobs().list(
+        projectId=PROJECT,
+        pageSize=50  # only return the 50 most recent results - our job is likely to be in here. If the job is not found, first try increasing this number
+    )
+    response = jobs_request.execute()
+
+    # search for the job in the list that has our name (names are unique)
+    for job in response['jobs']:
+        if job['name'] == job_name:
+            return job['id']
+
+    # if we don't find a job, just return
+    return
 
 
-def test_run_template_python_empty_args(app):
+# We retry the cancel operation a few times until the job is in a state where it can be cancelled
+@backoff.on_exception(backoff.expo, HttpError, max_time=240)
+def dataflow_jobs_cancel(job_name):
+    # to cancel a dataflow job, we need its ID, not its name
+    job_id = get_job_id_from_name(job_name)
+
+    if job_id:
+        # Cancel the Dataflow job if it exists. If it doesn't, job_id will be equal to None.
+        request = dataflow.projects().jobs().update(
```

Comment:
Nit: can we add a comment with a link to the docs as well?

https://cloud.google.com/dataflow/docs/reference/rest/v1b3/projects.jobs/update

Reply:
done